### PR TITLE
Add Rate Limit Support

### LINKED ISSUE #426: Add rate limit support
### Problem

The current implementation of the relay does now have application level rate limiting.

### Solution

Add application level rate limiting

- Explore [koajs rate limit](https://github.com/koajs/ratelimit) or other options
- Support method based rate limiting with a tighter limit on relay account costing methods
- Add support for account burn rate limiting

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by georgi-l95**
**Description**:
Adds application level rate limiting with support for method based rate limiting with a tighter limit on relay account costing methods. Custom solution implemented in `koa-jsonrpc` module.

**Related issue(s)**:

Fixes #426 

**Notes for reviewer**:
Some rate limit values in `methodConfiguration` module, perhaps should be modified.

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

**Review by Nana-EC — CHANGES_REQUESTED** [review]
Nice progress.
Please add documentation, `.env` level configuration and streamline the rate limiting logic removing parts that aren't necessary and capturing in a future ticket.
Also note HBAR comment

File: packages/server/src/koaJsonRpc/lib/RpcError.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  }
+}
+
+export class RateLimitExceeded extends JsonRpcError {
+  constructor() {
+    super('Rate limit exceeded', -32605, undefined);
+  }
+}
+
+// module.exports = (...args) => new JsonRpcError(...(args as [any, any, any]));
```
[the remark this answers is not in the document]
Comment (georgi-l95) [reply packages/server/src/koaJsonRpc/lib/RpcError.ts:93]: Done.

File: packages/server/src/koaJsonRpc/index.ts [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    this.registry = Object.create(null);
+    this.registryTotal = Object.create(null);
+    this.methodConfig = methodConfiguration;
+    if (opts) {
+      this.limit = opts.limit || this.limit;
+      this.auth = opts.auth;
+      this.duration = opts.limit || this.limit;
+    }
+    this.ratelimit = new RateLimit(this.duration);
+    if (this.auth && (!hasOwnProperty(this.auth, 'username') || !hasOwnProperty(this.auth, 'password'))) {
```
[the remark this answers is not in the document]
Comment (georgi-l95) [reply packages/server/src/koaJsonRpc/index.ts:59]: We don't have auth support, this is from the original `koa-jsonRpc` . Will remove it as we don't need it atm.

File: packages/server/src/koaJsonRpc/lib/methodConfiguration.ts [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    total: 400,
+  },
+  net_listening: {
+    total: 400,
+  },
+  net_version: {
+    total: 400,
+  },
+  eth_blockNumber: {
+    total: 200,
```
[the remark this answers is not in the document]
Comment (georgi-l95) [reply packages/server/src/koaJsonRpc/lib/methodConfiguration.ts]: Exposed them in the .env file, just added RATE to the name, to be more clear. ex `TIER_1_RATE_LIMIT` .

File: packages/server/src/koaJsonRpc/lib/methodConfiguration.ts [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+export const methodConfiguration = {
```
[the remark this answers is not in the document]
Comment (georgi-l95) [reply packages/server/src/koaJsonRpc/lib/methodConfiguration.ts:26]: Done.

File: packages/server/tests/acceptance/rpc.spec.ts [hunk 5]
```diff
@@ -1149,5 +1149,17 @@ describe('RPC Server Acceptance Tests', function () {
                 expect(Number(res.oldestBlock)).to.be.gt(0);
             });
         });
+
+        describe('RPC rate limit test', () => {
+            it('should throw rate limit exceeded error', async function() {
+                try{
+                    for (let index = 0; index < 201; index++) {
```
[the remark this answers is not in the document]
Comment (georgi-l95) [reply packages/server/tests/acceptance/rpc.spec.ts]: Done.

File: packages/server/src/ratelimit/index.ts [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+export default class RateLimit {
```
[the remark this answers is not in the document]
Comment (georgi-l95) [reply packages/server/src/ratelimit/index.ts:21]: Done, added.

File: packages/server/src/koaJsonRpc/index.ts [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+      }
+
+      if (!this.registry[body.method]) {
+        ctx.body = jsonResp(body.id, new MethodNotFound(), undefined);
+        return;
+      }
+
+      const methodName = body.method;
+      const methodTotalLimit = this.registryTotal[methodName];
+      if (this.ratelimit.shouldRateLimit(ctx.ip, methodName, methodTotalLimit)) {
```
[the remark this answers is not in the document]
Comment (georgi-l95) [reply packages/server/src/koaJsonRpc/index.ts:107]: Okay, can't think of a clear and easy way to implement this counter. If that's okay with you, can we open new issue to research and address this ?

[the remark this answers is not in the document]
Comment (Nana-EC) [reply packages/server/src/koaJsonRpc/index.ts:107]: Yes let's do that

**Comment by Nana-EC** [comment]
@rustyShacklefurd heads up as this contains some additional configs.
Defaults are probably okay but please advise on any concerns regarding ranges or deployment impacts

File: packages/server/src/server.ts [hunk 8]
```diff
@@ -39,13 +38,13 @@ const mainLogger = pino({
...
 });
+
 const logger = mainLogger.child({ name: 'rpc-server' });
 const register = new Registry();
-
 const relay: Relay = new RelayImpl(logger, register);
 const cors = require('koa-cors');
 const app = new Koa();
-const rpc = koaJsonRpc();
+const rpc = new KoaJsonRpc();
```
Comment (Nana-EC) [inline packages/server/src/server.ts:47]: Does using the custom implementation make any of the controller logic in this file moveable?
That is should anything here move to `koaJsonRPc/index.ts`? 
Don't have to do it in this PR but it's be good to know so we track that.

**Review by Nana-EC — APPROVED** [review]
LG

File: packages/server/src/koaJsonRpc/index.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (Nana-EC) [reply packages/server/src/koaJsonRpc/index.ts:107]: Added to https://github.com/hashgraph/hedera-json-rpc-relay/issues/528

File: packages/server/src/server.ts (hunk 8 above)
Comment (georgi-l95) [reply packages/server/src/server.ts:47]: I think that's doable. I've opened a separate issue to take care of this. Issue #530

**Pull request merged**